src/resources/wifi_esp32.cc

```diff
@@ -52,28 +52,59 @@ ResourcePool<int, kInvalidWifi> wifi_pool(
 class WifiResourceGroup : public ResourceGroup {
  public:
   TAG(WifiResourceGroup);
-  WifiResourceGroup(Process* process, SystemEventSource* event_source, int id,
-                    esp_netif_t* netif)
+  WifiResourceGroup(Process* process, SystemEventSource* event_source, int id, esp_netif_t* netif)
       : ResourceGroup(process, event_source)
       , _id(id)
       , _netif(netif) {
+    clear_ip_address();
   }
 
+  uint32 ip_address() const { return _ip_address; }
+  bool has_ip_address() const { return _ip_address != 0; }
+
+  void set_ip_address(uint32 address) { _ip_address = address; }
+  void clear_ip_address() { _ip_address = 0; }
+
   esp_err_t connect(const char* ssid, const char* password) {
-    wifi_config_t config;
-    memset(&config, 0, sizeof(config));
-    strncpy(char_cast(config.sta.ssid), ssid, sizeof(config.sta.ssid));
-    strncpy(char_cast(config.sta.password), password, sizeof(config.sta.password));
     // Configure the WiFi to _start_ the channel scan from the last connected channel.
     // If there has been no previous connection, then the channel is 0 which causes a normal scan.
     uint8 channel = RtcMemory::wifi_channel();
     if (!(0 <= channel && channel <= 13)) {
       channel = 0;
       RtcMemory::set_wifi_channel(0);
     }
+
+    esp_err_t err = esp_wifi_set_mode(WIFI_MODE_STA);
+    if (err != ESP_OK) return err;
+
+    wifi_config_t config;
+    memset(&config, 0, sizeof(config));
+    strncpy(char_cast(config.sta.ssid), ssid, sizeof(config.sta.ssid));
+    strncpy(char_cast(config.sta.password), password, sizeof(config.sta.password));
```

Comment:
I think we should be copying the min of the strlen and the size of the destination.
And perhaps zero the last byte of config.sta.password to ensure that it is null terminated for a too-long password.